File: app/src/app/api/kakao-route/route.ts
// 카카오내비 길찾기 프록시 — REST 키는 서버 전용 env(KAKAO_REST_API_KEY, NEXT_PUBLIC 아님).
// Render 프록시(api/routes/route.py)는 카카오가 Render egress IP를 플래그해 401(-401 KA)을
// 받으므로(2026-07-11 실측, GitHub 러너·가정망에서는 KA 없이도 성공) Vercel 함수로 이전.
const DIRECTIONS_URL = "https://apis-navi.kakaomobility.com/v1/directions";

function parseLatLng(value: string | null): [number, number] | null {
  if (!value) return null;
  const [lat, lng] = value.split(",").map(Number);
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null;
  return [lat, lng];
}

function asRecord(v: unknown): Record<string, unknown> | null {
  return typeof v === "object" && v !== null ? (v as Record<string, unknown>) : null;
}

export async function GET(request: Request) {
  const url = new URL(request.url);
  const from = parseLatLng(url.searchParams.get("from"));
  const to = parseLatLng(url.searchParams.get("to"));
  if (!from || !to) {
    return Response.json({ error: "from/to는 'lat,lng' 형식" }, { status: 400 });
  }
  const key = process.env.KAKAO_REST_API_KEY;
  if (!key) {
    return Response.json({ error: "KAKAO_REST_API_KEY 미설정" }, { status: 500 });
  }

  const qs = `origin=${from[1]},${from[0]}&destination=${to[1]},${to[0]}`;
  const res = await fetch(`${DIRECTIONS_URL}?${qs}`, {
    headers: { Authorization: `KakaoAK ${key}` },
  });
  if (!res.ok) {
    const body = await res.text();
    return Response.json(
      { error: `카카오내비 응답 ${res.status}: ${body.slice(0, 200)}` },
      { status: 502 },
    );
  }

  const payload: unknown = await res.json();
  const routes = asRecord(payload)?.routes;
  const route = asRecord(Array.isArray(routes) ? routes[0] : null);
  if (!route) return Response.json({ error: "경로 없음" }, { status: 404 });
  if (route.result_code !== 0) {
    // 103 등: 좌표가 도로에서 먼 경우 — 프론트는 외부 카카오맵 링크로 폴백
    return Response.json(
      { error: `경로 탐색 실패(${route.result_code}): ${route.result_msg}` },
      { status: 422 },
    );
  }

  const summary = asRecord(route.summary);
  const distance = summary?.distance;
  const duration = summary?.duration;
  if (typeof distance !== "number" || typeof duration !== "number") {
    return Response.json({ error: "summary 형식 불량" }, { status: 502 });
  }

  const path: [number, number][] = [];
  const sections = Array.isArray(route.sections) ? route.sections : [];
  for (const secRaw of sections) {
    const roads = asRecord(secRaw)?.roads;
    for (const roadRaw of Array.isArray(roads) ? roads : []) {
      const vertexes = asRecord(roadRaw)?.vertexes;
      if (!Array.isArray(vertexes)) continue;
      // vertexes는 [x1, y1, x2, y2, ...] 평탄 배열 (x=lng, y=lat)
      for (let i = 0; i < vertexes.length - 1; i += 2) {
        const x = vertexes[i];
        const y = vertexes[i + 1];
        if (typeof x === "number" && typeof y === "number") path.push([y, x]);
      }
    }
  }
  if (path.length === 0) {
    return Response.json({ error: "경로 좌표 없음" }, { status: 502 });
  }
  return Response.json({ distance_m: distance, duration_s: duration, path });
}

File: api/routes/route.py
"""GET /route — 일정 슬롯 간 자동차 경로 (카카오내비 길찾기 API 프록시).

REST 키는 서버 전용이므로 프론트가 직접 호출하지 않고 이 엔드포인트를 거친다.
응답 path는 [lat, lng] 좌표열 — 프론트가 카카오맵 Polyline으로 그린다.
"""

from __future__ import annotations

from functools import lru_cache

import requests
from fastapi import APIRouter, HTTPException

from api.core.http import ExternalApiError, as_dict, as_list, get_json
from api.deps import get_settings, get_spot_meta
from api.schemas import RouteOut

router = APIRouter()

DIRECTIONS_URL = "https://apis-navi.kakaomobility.com/v1/directions"


@lru_cache(maxsize=1)
def _session() -> requests.Session:
    s = requests.Session()
    s.headers["Authorization"] = f"KakaoAK {get_settings().kakao_rest_api_key}"
    # 데이터센터발 요청은 KA 헤더 없이 401로 거부됨 ("KA Header is required" — Render 실측)
    s.headers["KA"] = "sdk/1.0 os/rest lang/ko origin/https://jejunow.vercel.app"
    return s


@router.get("/route", response_model=RouteOut)
def get_route(from_spot: int, to_spot: int) -> RouteOut:
    spots = get_spot_meta()
    origin, dest = spots.get(from_spot), spots.get(to_spot)
    if origin is None or dest is None:
        raise HTTPException(status_code=404, detail=f"미등록 spot_id={from_spot or to_spot}")
    params = {
        "origin": f"{origin.lng},{origin.lat}",
        "destination": f"{dest.lng},{dest.lat}",
    }
    try:
        payload = get_json(_session(), DIRECTIONS_URL, params)
    except ExternalApiError as exc:
        raise HTTPException(status_code=502, detail=f"길찾기 API 실패: {exc}") from exc

    routes = as_list(as_dict(payload, "directions").get("routes", []), "routes")
    if not routes:
        raise HTTPException(status_code=404, detail="경로 없음")
    route = as_dict(routes[0], "route")
    code = route.get("result_code")
    if code != 0:
        # 103 등: 스팟 좌표가 도로에서 먼 경우 — 프론트는 외부 카카오맵 링크로 폴백
        raise HTTPException(
            status_code=422, detail=f"경로 탐색 실패({code}): {route.get('result_msg')}"
        )

    summary = as_dict(route.get("summary", {}), "summary")
    distance, duration = summary.get("distance"), summary.get("duration")
    if not isinstance(distance, int) or not isinstance(duration, int):
        raise HTTPException(status_code=502, detail="summary 형식 불량")

    path: list[tuple[float, float]] = []
    for sec_obj in as_list(route.get("sections", []), "sections"):
        sec = as_dict(sec_obj, "section")
        for road_obj in as_list(sec.get("roads", []), "roads"):
            road = as_dict(road_obj, "road")
            vertexes = as_list(road.get("vertexes", []), "vertexes")
            # vertexes는 [x1, y1, x2, y2, ...] 평탄 배열 (x=lng, y=lat)
            for i in range(0, len(vertexes) - 1, 2):
                x, y = vertexes[i], vertexes[i + 1]
                if isinstance(x, (int, float)) and isinstance(y, (int, float)):
                    path.append((float(y), float(x)))
    if not path:
        raise HTTPException(status_code=502, detail="경로 좌표 없음")
    return RouteOut(distance_m=distance, duration_s=duration, path=path)
